crates/astria-celestia-client/src/blob_space.rs

```diff
@@ -0,0 +1,320 @@
+use sequencer_validation::InclusionProof;
+use serde::{
+    de::DeserializeOwned,
+    Deserialize,
+    Serialize,
+};
+use tendermint::{
+    block::{
+        Commit,
+        Header,
+    },
+    Hash,
+};
+
+/// Utility to create a v0 celestia namespace from the sha256 of `bytes`.
+#[must_use]
+pub fn celestia_namespace_v0_from_hashed_bytes(bytes: &[u8]) -> Namespace {
+    use sha2::{
+        Digest as _,
+        Sha256,
+    };
+    let mut hasher = Sha256::new();
+    hasher.update(bytes);
+    let result = hasher.finalize();
+    const _: () = assert!(NS_ID_V0_SIZE < 32);
+    let id = <[u8; NS_ID_V0_SIZE]>::try_from(&result[0..NS_ID_V0_SIZE])
+        .expect("must not fail as hash is always 32 bytes and NS_ID_V0_SIZE < 32");
+    Namespace::const_v0(id)
+}
+
+fn hash_json_serialized_bytes<T: Serialize>(val: &T) -> [u8; 32] {
+    use sha2::{
+        Digest,
+        Sha256,
+    };
+    let mut hasher = Sha256::new();
+    let bytes = serde_json::to_vec(val).expect(
+        "should not fail because all types called with this do not contain maps and hence \
+         non-unicode keys that would trigger to_vec()'s only error case",
+    );
+    hasher.update(bytes);
+    hasher.finalize().into()
+}
+
+/// A wrapper of some abstract namespace data together with the public key
+/// derived from the signing key that created the signature.
+#[derive(Clone, Serialize, Deserialize, Debug)]
+#[serde(try_from = "UnverifiedSignedNamespaceData::<D>")]
+#[serde(into = "UnverifiedSignedNamespaceData::<D>")]
+#[serde(bound = "D: Clone + DeserializeOwned + Serialize")]
+pub struct SignedNamespaceData<D> {
+    data: D,
+    public_key: VerificationKey,
+    signature: Signature,
+}
+
+impl<D> SignedNamespaceData<D> {
+    pub fn data(&self) -> &D {
+        &self.data
+    }
+
+    pub fn public_key(&self) -> VerificationKey {
+        self.public_key
+    }
+
+    pub fn signature(&self) -> Signature {
+        self.signature
+    }
+
+    pub fn from_unverified_unchecked(unverified: UnverifiedSignedNamespaceData<D>) -> Self {
+        let UnverifiedSignedNamespaceData {
+            data,
+            public_key,
+            signature,
+        } = unverified;
+        Self {
+            data,
+            public_key,
+            signature,
+        }
+    }
+
+    pub fn into_unverified(self) -> UnverifiedSignedNamespaceData<D> {
+        let Self {
+            data,
+            public_key,
+            signature,
+        } = self;
+        UnverifiedSignedNamespaceData {
+            data,
+            public_key,
+            signature,
+        }
+    }
+}
+
+impl<D: Serialize> SignedNamespaceData<D> {
+    /// Constructs a `SignedNamespaceData` by signing the sha256 hashed JSON serialization
+    /// of `data`.
+    ///
+    /// NOTE: Json is unstable under serialization. This must be replaced ASAP.
+    pub fn from_data_and_key(data: D, key: &SigningKey) -> Self {
+        let hash = hash_json_serialized_bytes(&data);
```

Comment:
do we want this to be json still or are you going to update this in another PR?

Reply:
Yeah, we don't want it to be json. But moving it all to protobuf is a lot of work and outside the scope of this PR. #12 and #443